doc/arch/013-sql-performance-profiling-v2.md

```diff
@@ -0,0 +1,104 @@
+
+```java
+private static <T> T measureAndReport(final String methodName, final kotlin.jvm.functions.Function0<T> block) {
+    final long start = System.currentTimeMillis();
+    final T result;
+    result = block.invoke();
+    final java.time.Duration timeTaken = java.time.Duration.ofMillis(System.currentTimeMillis() - start);
+    org.simple.clinic.SqlPerformanceReporter.report("UserRoomDao_Impl", methodName, timeTaken);
+    return result;
+ }
+```
+
+This would allow us to define a class, `org.simple.clinic.SqlPerformanceReporter`, whose `report` method would be called everytime the `measureAndReport` method would be invoked.
+
+#### Step 2: Delegate DAO query operation
+The next step would then be to delegate the actual SQL query operations in every method in the DAO to the `measureAndReport` method. Considering a generated method like
+
+
+```java
+@Override
+public UUID currentFacilityUuid() {
+    final String _sql = "SELECT currentFacilityUuid FROM LoggedInUser LIMIT 1";
+    final RoomSQLiteQuery _statement = RoomSQLiteQuery.acquire(_sql, 0);
+    __db.assertNotSuspendingTransaction();
+    final Cursor _cursor = DBUtil.query(__db, _statement, false, null);
+    try {
+      final UUID _result;
+      if(_cursor.moveToFirst()) {
+        final String _tmp;
+        _tmp = _cursor.getString(0);
+        _result = __uuidRoomTypeConverter.toUuid(_tmp);
+      } else {
+        _result = null;
+      }
+      return _result;
+    } finally {
+      _cursor.close();
+      _statement.release();
+    }
+}
+```
+
+We then transformed this method so that it looked like
+
+```java
+@Override
+public UUID currentFacilityUuid() {
+    return measureAndReport("currentFacilityUuid", () -> {
+      final String _sql = "SELECT currentFacilityUuid FROM LoggedInUser LIMIT 1";
+      final RoomSQLiteQuery _statement = RoomSQLiteQuery.acquire(_sql, 0);
+      __db.assertNotSuspendingTransaction();
+      final Cursor _cursor = DBUtil.query(__db, _statement, false, null);
+      try {
+        final UUID _result;
+        if (_cursor.moveToFirst()) {
+          final String _tmp;
+          _tmp = _cursor.getString(0);
+          _result = __uuidRoomTypeConverter.toUuid(_tmp);
+        } else {
+          _result = null;
+        }
+        return _result;
+      } finally {
+        _cursor.close();
+        _statement.release();
+      }
+    });
+}
+```
+
+Adding this step to the build allowed us to automatically capture SQL performance profiling at whatever level of granularity we need.
+
+### Implementation
+The implementation of the build tool that transforms the generated Room DAOs is available [here](https://github.com/simpledotorg/room-metadata-generator). The integration with the `simple-android` build process and the reporting of the events to our analytics platform is available [here](https://github.com/simpledotorg/simple-android/pull/2703).
+
+## Consequences
```

Comment:
Should we add limitations like paging again here?

Reply:
This is no longer a limitation since we can add support for it. The older method made it impossible.